## Computational Assignment 2: Statistical Inference in Linear Regression

### MSDS 410

In [3]:
import numpy as np
from scipy import stats

# ----- Model 1: SalePrice ~ X1 + X2 + X3 + X4 -----
model1 = {
    'n_predictors': 4,
    'coef': {
        'Intercept': {'est': 11.3303, 'se': 1.9941, 't': 5.68,  'p': 0.0001},
        'X1': {'est': 2.186,   'se': 0.4104, 't': None,  'p': 0.0001},  
        'X2': {'est': 8.2743,  'se': 2.3391, 't': 3.54,  'p': 0.0007},
        'X3': {'est': 0.49182, 'se': 0.2647, 't': 1.86,  'p': 0.0676},
        'X4': {'est': -0.49356,'se': 2.2943, 't': -0.22, 'p': 0.8303},
    },
    'anova': {
        'X1': {'df': 1,  'ss': 1974.53, 'ms': 1974.53, 'F': 209.8340, 'p': 0.0001},
        'X2': {'df': 1,  'ss': 118.8642568, 'ms': 118.8642568, 'F': 12.6339, 'p': 0.0007},
        'X3': {'df': 1,  'ss': 32.47012585, 'ms': 32.47012585, 'F': 3.4512, 'p': 0.0676},
        'X4': {'df': 1,  'ss': 0.435606985, 'ms': 0.435606985, 'F': 0.0463, 'p': 0.8303},
        'Residuals': {'df': 67, 'ss': 630.36, 'ms': 9.41},
        'Model': {'df': 4, 'ss': 2126, 'ms': 531.50, 'p': 0.0001}, 
        'Total': {'df': 71, 'ss': 2756.37},
    },
    'resid_se': 3.06730,
    'mult_r2': 0.7713,
    'adj_r2': 0.7577,
}

# ----- Model 2: SalePrice ~ X1 + X2 + X3 + X4 + X5 + X6 -----
model2 = {
    'n_predictors': 6,
    'coef': {
        'Intercept': {'est': 14.3902, 'se': 2.89157, 't': 4.98, 'p': 0.0001},
        'X1': {'est': 1.97132, 'se': 0.43653, 't': 4.52, 'p': 0.0001},
        'X2': {'est': 9.13895, 'se': 2.30071, 't': 3.97, 'p': 0.0002},
        'X3': {'est': 0.56485, 'se': 0.26266, 't': 2.15, 'p': 0.0352},
        'X4': {'est': 0.33371, 'se': 2.42131, 't': 0.14, 'p': 0.8908},
        'X5': {'est': 1.90698, 'se': 0.76459, 't': 2.49, 'p': 0.0152},
        'X6': {'est': -1.0433, 'se': 0.64759, 't': -1.61, 'p': 0.1120},
    },
    'anova': {
        'X1': {'df': 1, 'ss': 1928.27000, 'ms': 1928.27000, 'F': 218.8890, 'p': 0.0001},
        'X2': {'df': 1, 'ss': 136.92075, 'ms': 136.92075, 'F': 15.5426, 'p': 0.0002},
        'X3': {'df': 1, 'ss': 40.75872, 'ms': 40.75872, 'F': 4.6267, 'p': 0.0352},
        'X4': {'df': 1, 'ss': 0.16736, 'ms': 0.16736, 'F': 0.0190, 'p': 0.8908},
        'X5': {'df': 1, 'ss': 54.77667, 'ms': 54.77667, 'F': 6.2180, 'p': 0.0152},
        'X6': {'df': 1, 'ss': 22.86647, 'ms': 22.86647, 'F': 2.5957, 'p': 0.1120},
        'Residuals': {'df': 65, 'ss': 572.60910,  'ms': 8.80937},
        'Model': {'df': 6, 'ss': 2183.75946, 'ms': 363.96, 'F': 41.3200, 'p': 0.0001},
        'Total': {'df': 71, 'ss': 2756.37},
    },
    'resid_se': 2.968,
    'mult_r2': 0.7923,
    'adj_r2': 0.7731,
}

print("Data loaded for Model 1 and Model 2.")

Data loaded for Model 1 and Model 2.


---
### Model 1

#### Question 1

**How many observations are in the sample data?**

The number of observations can be recovered from the ANOVA table's Total degrees of freedom: 
$$n-1 = df_{Total} => df_{Total} + 1$$

In [6]:
df_total = model1['anova']['Total']['df']
n = df_total + 1
print(f"df_Total = {df_total}")
print(f"n = df_Total + 1 = {n}")

df_Total = 71
n = df_Total + 1 = 72


There are n = 72 observations in the sample data used to fit Model 1

#### Question 2

**Write out the null and alternate hypotheses for the t-test for Beta1.**

Test of whether $X_1$ has a significant linear relationship with the response, holding the other predictors constant: 
$$H_0 = \beta_1 = 0$$
$$H_a : \beta_1 \neq 0$$

#### Question 3

**Compute the t-statistic for Beta1. Conduct the hypothesis test and interpret the result.**

$$t = \frac{\hat \beta_1 - 0}{SE(\hat \beta_1)}$$

with $df = n - k - 1$, where $k$ = number of predictors in Model 1 (4). 

In [13]:
est_X1 = model1['coef']['X1']['est']
se_X1 = model1['coef']['X1']['se']
k1 = model1['n_predictors']
df_resid1 = n - k1 - 1

t_X1 = (est_X1 - 0) / se_X1
p_X1 = 2 * (1 - stats.t.cdf(abs(t_X1), df_resid1))
t_crit = stats.t.ppf(0.975, df_resid1)

print(f"df (residual) = {df_resid1} (matches ANOVA Residuals df = {model1['anova']['Residuals']['df']})")
print(f"t = {est_X1} / {se_X1} = {t_X1:.4f}")
print(f"two-sided p-value = {p_X1:.6f}")
print(f"critical t (alpha=0.05, df={df_resid1}) = +-{t_crit:.4f}")

df (residual) = 67 (matches ANOVA Residuals df = 67)
t = 2.186 / 0.4104 = 5.3265
two-sided p-value = 0.000001
critical t (alpha=0.05, df=67) = +-1.9960


$t = 5.3265$, with $df = 67$. Since $|t| = 5.3265 > t_{crit} = 1.9960$ (p < 0.0001), we reject $H_0$. There is statistically significant evidence that $\beta \neq 0 - X_1$ has a significant linear relationship with the response, holding $X_2$, $X_3$, $X_4$ constant. 

#### Question 4

**Compute the R-Squared value for Model 1, using information from the ANOVA table. Interpret this statistic.**
$$R^2 = \frac{SS_{Model}}{SS_{Total}} = 1 - (\frac{SS_{Residual}}{SS_{Total}})$$

In [20]:
SS_model1 = model1['anova']['Model']['ss']
SS_resid1 = model1['anova']['Residuals']['ss']
SS_total = model1['anova']['Total']['ss']

R2_1 = SS_model1 / SS_total
R2_1_check = 1 - SS_resid1/SS_total

print(f"R^2 = {SS_model1} / {SS_total} = {R2_1:.4f}")
print(f"check (1 - SSE/SST) = {R2_1_check:.4f}")
print(f"Reported in R output: {model1['mult_r2']}")

R^2 = 2126 / 2756.37 = 0.7713
check (1 - SSE/SST) = 0.7713
Reported in R output: 0.7713


$R^2 = 0.7713$. Approximately 77.13% of the variability in the response is explained by the four predictors ($X_1 - X_4$) jointly in Model 1. 

#### Question 5

**Compute the Adjusted R-Squared value for Model 1. Discuss why Adjusted R-squared and the R-squared value are different.**
$$R_{adj}^2 = 1 - (1 - R^2) \frac{n-1}{n-k-1}$$

In [23]:
k1 = model1['n_predictors']
adjR2_1 = 1 - (1 - R2_1) * (n - 1) / (n - k1 - 1)

print(f"n = {n}, k = {k1}")
print(f"Adjusted R^2 = 1 - (1-{R2_1:.4f}) * ({n}-1)/({n}-{k1}-1) = {adjR2_1:.4f}")
print(f"Reported in R output: {model1['adj_r2']}")

n = 72, k = 4
Adjusted R^2 = 1 - (1-0.7713) * (72-1)/(72-4-1) = 0.7577
Reported in R output: 0.7577


Adjusted $R^2 = 0.7577$, slightly lower than $R^2 = 0.7713$. Ordinary $R^2$ increases every time an additional predictor is added to the model, regardless of whether that predictor is actually useful. Adjusted $R^2$ imposes a penalty for the number of predictors relative to the sample size, so it only increases if a new predictor actually improves the fit by more than what would be expected by chance. This makes Adjusted $R^2$ more appropriate for comparing models with different numbers of predictors. 

#### Question 6

**Write out the null and alternate hypothesis for the Overall F-test.**

$$H_0 = \beta_1 = \beta_2 = \beta_3 = \beta_4 = 0$$
$$H_a = \text{at least one } \beta_j \neq 0, j = 1,2,3,4$$

#### Question 7

**Compute the F-statistic for the Overall F-test. Conduct the hypothesis test and interpret the result.**

$$F = \frac{MS_{Model}}{MS_{Residual}} = \frac{SS_{Model}/df_{Model}}{SS_{Residual}/df_{Residual}}$$

In [27]:
MS_model1 = model1['anova']['Model']['ms']
MS_resid1 = model1['anova']['Residuals']['ms']
df_model1 = model1['anova']['Model']['df']
df_resid1 = model1['anova']['Residuals']['df']

F1 = MS_model1 / MS_resid1
p_F1 = 1 - stats.f.cdf(F1, df_model1, df_resid1)
F1_crit = stats.f.ppf(0.95, df_model1, df_resid1)

print(f"F = {MS_model1} / {MS_resid1} = {F1:.4f}")
print(f"df = ({df_model1}, {df_resid1})")
print(f"p-value = {p_F1:.6f}")
print(f"F critical (alpha=0.05) = {F1_crit:.4f}")

F = 531.5 / 9.41 = 56.4825
df = (4, 67)
p-value = 0.000000
F critical (alpha=0.05) = 2.5087


$F = 56.4825$ with $df = (4,67)$. Since $F = 56.4825 > F_{crit} = 2.5087$ (p < 0.0001), we reject $H_0$. The overall regression is statistically significant. The four combined predictors explain a significant amount of the variability in the response. 

---
### Model 1 vs. Model 2

#### Question 8

**Does Model 1 nest Model 2 or does Model 2 nest Model 1? Explain.**
- Model 1: response $~ X_1 + X_2 + X_3 + X_4$
- Model 2: response $~ X_1 + X_2 + X_3 + X_4 + X_5 + X_6$

Model 1's predictors are a subset of Model 2's predictors. Model 1 can be obtained from Model 2 by restricting $\beta_5 = \beta_6 = 0$. Model 2 nests Model 1. 

#### Question 9

**Write out the null and alternate hypotheses for nested F-test using Model 1 and Model 2.**
$$H_0 = \beta_5 = \beta_6 = 0$$
$$H_a = \text{at least one of } \beta_5, \beta_6 \neq 0$$

#### Question 10

**Compute the F-statistic for a nested F-test using Model 1 and Model 2. Conduct the hypothesis test and interpret the results.**
$$F = \frac{(SSE{reduced} - SSE{full})/(df_{reduced} - df_{full})}{SSE_{full}/df_{full}}$$

where "reduced" = Model 1 and "full" = Model 2

In [36]:
SSE_reduced = model1['anova']['Residuals']['ss']
df_reduced = model1['anova']['Residuals']['df']

SSE_full = model2['anova']['Residuals']['ss']
df_full = model2['anova']['Residuals']['df']

df_num = df_reduced - df_full
F_nested = ((SSE_reduced - SSE_full) / df_num) / (SSE_full / df_full)
p_nested = 1 - stats.f.cdf(F_nested, df_num, df_full)
F_nested_crit = stats.f.ppf(0.95, df_num, df_full)

print(f"SSE_reduced (Model 1) = {SSE_reduced}, df = {df_reduced}")
print(f"SSE_full (Model 2) = {SSE_full}, df = {df_full}")
print(f"numerator df = {df_reduced} - {df_full} = {df_num}")
print()
print(f"F = [({SSE_reduced} - {SSE_full}) / {df_num}] / [{SSE_full} / {df_full}]")
print(f"F = {F_nested:.4f}")
print(f"df = ({df_num}, {df_full})")
print(f"p-value = {p_nested:.4f}")
print(f"F critical (alpha=0.05) = {F_nested_crit:.4f}")

SSE_reduced (Model 1) = 630.36, df = 67
SSE_full (Model 2) = 572.6091, df = 65
numerator df = 67 - 65 = 2

F = [(630.36 - 572.6091) / 2] / [572.6091 / 65]
F = 3.2778
df = (2, 65)
p-value = 0.0440
F critical (alpha=0.05) = 3.1381


$F = 3.2778$ with $df = (2,65), p = 0.0440$. Since $F = 3.2778 > F_{crit} = 3.1381$, we reject $H_0$. Adding $X_5$ and $X_6$ to Model 1 produces a statistically significant improvement in fit at the 95% confidence level, so Model 2 is preferred over Model 1. 